src/full/Agda/TypeChecking/RecordPatterns.hs

```diff
@@ -60,29 +60,39 @@ import Agda.Utils.Impossible
 --
 --   E.g. for @(x , (y , z))@ we return @[ fst, fst . snd, snd . snd ]@.
 --
---   If it is not a record pattern, error 'ShouldBeRecordPattern' is raised.
+--   If it is not a record pattern, error 'ShouldBeEtaRecordPattern' is raised.
 recordPatternToProjections :: DeBruijnPattern -> TCM [Term -> Term]
 recordPatternToProjections p =
   case p of
     VarP{}       -> return [ id ]
-    LitP{}       -> typeError $ ShouldBeRecordPattern p
-    DotP{}       -> typeError $ ShouldBeRecordPattern p
+    LitP{}       -> impossible "LitP"
+
+    DotP{}       -> impossible "DotP"
+    IApplyP{}    -> impossible "IApplyP"
+    DefP{}       -> impossible "DefP"
+    ProjP{}      -> impossible "ProjP"
+
     ConP c ci ps -> do
       unless (conPRecord ci) $
-        typeError $ ShouldBeRecordPattern p
-      let t = unArg $ fromMaybe __IMPOSSIBLE__ $ conPType ci
+        typeError $ ShouldBeEtaRecordPattern DataNotRecord p
+      t <- reduce (unArg $ fromMaybe __IMPOSSIBLE__ $ conPType ci)
       reportSDoc "tc.rec" 45 $ vcat
         [ "recordPatternToProjections: "
         , nest 2 $ "constructor pattern " <+> prettyTCM p <+> " has type " <+> prettyTCM t
         ]
       reportSLn "tc.rec" 70 $ "  type raw: " ++ show t
-      fields <- getRecordTypeFields t
-      concat <$> zipWithM comb (map proj fields) (map namedArg ps)
-    ProjP{}      -> __IMPOSSIBLE__ -- copattern cannot appear here
-    IApplyP{}    -> typeError $ ShouldBeRecordPattern p
-    DefP{}       -> typeError $ ShouldBeRecordPattern p
+      case unEl t of
+        Def r _ -> fmap theDef (getConstInfo r) >>= \case
+          rt@Record { recFields = fields } | YesEta == recEtaEquality rt -> do
+            fields <- getRecordTypeFields t
```

Comment:
You meant to delete this call to `getRecordTypeFields`, I suppose.

Reply:
Woops, you're right.